src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceResponseTracker.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.GrainDirectory;
+using Orleans.SystemTargetInterfaces;
+
+namespace Orleans.Runtime.GrainDirectory 
+{
+    /// <summary>
+    /// A class that encapsulates response processing logic.
+    /// It is a promise that fires once it has enough responses to make a determination.
+    /// </summary>
+    internal class GlobalSingleInstanceResponseTracker : TaskCompletionSource<GlobalSingleInstanceResponseTracker.Outcome> {
```

Comment:
not sure why it inherits from TaskCompletionSource. What is it actually tracking? the array of responses is assigned and never changed, so what would make the task complete later if it wasn't already completed? 
Actually I now see that GSIRegistrar.SendRequest modifies the array after the fact. This could be very tricky to troubleshoot. Did you consider having a tracker object that tracks a list of Task<RemoteClusterActivationResponse>... basically instead of a TaskCompletionSource, the tracker exposes Task.WhenAll(indidualRequests). This way it would be easier to follow what the code is doing, and you also get rid of the need to call Notify() all the time just in case this was the last request to complete.